cpp/src/fil/fil.cu

```diff
@@ -382,47 +409,55 @@ void node2fil_dense(std::vector<dense_node_t>* pnodes, int root, int cur,
   node2fil_dense(pnodes, root, left + 1, tree, tl_node_at(tree, tl_right));
 }
 
-void node2fil_sparse(std::vector<sparse_node_t>* pnodes, int root, int cur,
-                     const tl::Tree& tree, const tl::Tree::Node& node) {
-  if (node.is_leaf()) {
-    sparse_node_init(&(*pnodes)[root + cur], node.leaf_value(), 0, 0, false,
-                     true, 0);
-    return;
-  }
-
-  // inner node
-  ASSERT(node.split_type() == tl::SplitFeatureType::kNumerical,
-         "only numerical split nodes are supported");
-  // tl_left and tl_right are indices of the children in the treelite tree
-  // (stored  as an array of nodes)
-  int tl_left = node.cleft(), tl_right = node.cright();
-  bool default_left = node.default_left();
-  float threshold = node.threshold();
-  adjust_threshold(&threshold, &tl_left, &tl_right, &default_left, node);
-
-  // reserve space for child nodes
-  // left is the offset of the left child node relative to the tree root
-  // in the array of all nodes of the FIL sparse forest
-  int left = pnodes->size() - root;
-  pnodes->push_back(sparse_node_t());
-  pnodes->push_back(sparse_node_t());
-  sparse_node_init(&(*pnodes)[root + cur], 0, threshold, node.split_index(),
-                   default_left, false, left);
-
-  // init child nodes
-  node2fil_sparse(pnodes, root, left, tree, tl_node_at(tree, tl_left));
-  node2fil_sparse(pnodes, root, left + 1, tree, tl_node_at(tree, tl_right));
-}
-
 void tree2fil_dense(std::vector<dense_node_t>* pnodes, int root,
                     const tl::Tree& tree) {
   node2fil_dense(pnodes, root, 0, tree, tl_node_at(tree, tree_root(tree)));
 }
 
 int tree2fil_sparse(std::vector<sparse_node_t>* pnodes, const tl::Tree& tree) {
+  typedef std::pair<const tl::Tree::Node*, int> pair_t;
+  std::stack<pair_t> stack;
   int root = pnodes->size();
   pnodes->push_back(sparse_node_t());
-  node2fil_sparse(pnodes, root, 0, tree, tl_node_at(tree, tree_root(tree)));
+  stack.push(pair_t(&tl_node_at(tree, tree_root(tree)), 0));
+  while (!stack.empty()) {
```

Comment:
Are these two recursion->stack conversions the main contributor to the speedups? They look correct to me, but they unfortunately increase the code complexity by a fair bit. It's particularly noticeable in this one.

Reply:
looks like ~30 lines get replaced by ~40. As someone who is trying to add ~1K lines, I cannot claim this is too complicated. if readability is a concern, we could consider refactoring into a treelite traversal function, which can accept two lambdas - to convert branch and leaf nodes.